### ISSUE TITLE
Add the possibility to fetch using Specification and Pageable but without doing a count query

### ISSUE THREAD

**Issue by Artur-** [spring-projects/spring-data-jpa#2762]
If you use `Page<T> findAll(Pageable pageable)` then Spring Data will execute the select plus the count query for every call. To get rid of the count query when you do not need the total count, you can add a find by without any filters: `List<T> findBy(Pageable pageable)`. You cannot add a `List<T> findAll(Pageable pageable)` as it will conflict with `findAll` that returns a `Page<T>`.

However when you use a `Specification` for filtering you still would want the same possibility: to get rid of the count query.
The available findAll is `Page<T> findAll(Specification<T> spec, Pageable pageable)` which again prevents adding a `List<T> findAll(Specification<T> spec, Pageable pageable)`. Also it seems that the `findBy` "hack" cannot be used as a method like `List<T> findBy(Specification<T> filter, Pageable pageable);` will throw
```
org.springframework.dao.InvalidDataAccessApiUsageException: At least 1 parameter(s) provided but only 0 parameter(s) present in query
```

Is there another way to accomplish this or could it be added as a feature?

**Comment by christophstrobl** [comment]
We could enhance the `FetchableFluentQuery` API to cover that usecase returning a `Slice` instead of a `Page`.

**Comment by Dylan0405** [comment]
could anybody tell me that has this problem solved or how could I realize this feature?

**Comment by Dylan0405** [comment]
yeah, finally, I also ended up with  a custom repository method implemention. 
Just like [https://gist.github.com/tcollins/0ebd1dfa78028ecdef0b](url). displayed.
`    @Override
    public Slice<T> findAllSliced(Specification<T> spec, Pageable pageable) {
        TypedQuery<T> query = getQuery(spec, pageable.getSort());

        query.setFirstResult((int) pageable.getOffset());
        int extraSize = pageable.getPageSize() + 1;
        query.setMaxResults(extraSize);

        List<T> result = query.getResultList();
        boolean hasNext = result.size() == extraSize;

        if(hasNext){
            result.remove(extraSize - 1);
        }
        return new SliceImpl<>(result, pageable, hasNext);
    }`

**Comment by membersound** [comment]
Please also cover this for `QuerydslPredicateExecutor.findBy(Predicate predicate, Function<FluentQuery.FetchableFluentQuery<S>, R> queryFunction)`.

**Comment by ManeeshSharma17** [comment]
> yeah, finally, I also ended up with a custom repository method implemention. Just like [https://gist.github.com/tcollins/0ebd1dfa78028ecdef0b](url). displayed. ` @OverRide public Slice findAllSliced(Specification spec, Pageable pageable) { TypedQuery query = getQuery(spec, pageable.getSort());
> 
> ```
>     query.setFirstResult((int) pageable.getOffset());
>     int extraSize = pageable.getPageSize() + 1;
>     query.setMaxResults(extraSize);
> 
>     List<T> result = query.getResultList();
>     boolean hasNext = result.size() == extraSize;
> 
>     if(hasNext){
>         result.remove(extraSize - 1);
>     }
>     return new SliceImpl<>(result, pageable, hasNext);
> }`
> ```

This is escaping some records in the result set

**Comment by kmeyer-mbs** [comment]
> We could enhance the `FetchableFluentQuery` API to cover that usecase returning a `Slice` instead of a `Page`.

Any updates on this? Thanks.

**Comment by max-buildops** [comment]
If you use MySQL (other DBs might have it as well), your second query could be a super cheap one: `SELECT FOUND_ROWS();`
Assuming your original statement has `LIMIT` as part of your select (`SELECT ... LIMIT pageSize`), above subsequent query returns a number of rows that would be returned if there was no `LIMIT` clause.
https://dev.mysql.com/doc/refman/8.4/en/information-functions.html#function_found-rows

Hope this helps.

**Comment by membersound** [comment]
@ Spring-Team: could the `found_rows()` function be a better alternative to the implicit executed `count()`? Maybe by activation with a parameter on the Repository or class-method?

**Comment by mp911de** [comment]
Fixed via #3727

**Comment by torshid** [comment]
> Fixed via [#3727](https://github.com/spring-projects/spring-data-jpa/pull/3727)

@mp911de How does this fix the issue? The documentation does not state that providing a null countSpec will prevent making the count query.

**Comment by mp911de** [comment]
#3727 has introduced dedicated support to return a `Slice`. Changes in commons see https://github.com/spring-projects/spring-data-commons/blob/4.0.x/src/main/java/org/springframework/data/repository/query/FluentQuery.java#L223

This ticket discusses various things from various angles and the net result is that we accept a count `Specification` and have implemented native `Slice` support.
